### ISSUE TITLE
CRFasRNN training not improving

### ISSUE THREAD

**Issue by annkitkat** [torrvision/crfasrnn#55]
Hi, 

I am pretty new to caffe in general and I try to train your CRFasRNN network using the PASCAL VOC data of 2012 in order to get an insight how this works and how the data needs to be prepared. Later I would like to use only some parts of your work and combine it with other layers. 

However, I used the solver and train prototxt file in the examples folder under segmentationcrfasrnn but the accuracy is pretty jumpy and low. The training has reached the 19,800th iteration and the output for the test accuracy at each 1000th iteration is the following:

Test net output #0: accuracy = 0.0778186
Test net output #0: accuracy = 0.028924
Test net output #0: accuracy = 0.0566633
Test net output #0: accuracy = 0.0345619
Test net output #0: accuracy = 0.0419963
Test net output #0: accuracy = 0.0290244
Test net output #0: accuracy = 0.0300839
Test net output #0: accuracy = 0.0460825
Test net output #0: accuracy = 0.0236691
Test net output #0: accuracy = 0.0320318
Test net output #0: accuracy = 0.0268505
Test net output #0: accuracy = 0.029318
Test net output #0: accuracy = 0.042781
Test net output #0: accuracy = 0.0517754
Test net output #0: accuracy = 0.0374599
Test net output #0: accuracy = 0.0413573
Test net output #0: accuracy = 0.0380903
Test net output #0: accuracy = 0.0351139
Test net output #0: accuracy = 0.0313849
Test net output #0: accuracy = 0.0325979

As far as I understand the train accuracy and loss are not conclusive here because of batch size = 1. 

Could anybody help me figure out how to fix this? 

Thanks a lot!
Annika

----
The steps I did so far: 

1. Prepare my data
- download PASCAL data from [http://host.robots.ox.ac.uk/pascal/VOC/voc2012/#devkit](http://host.robots.ox.ac.uk/pascal/VOC/voc2012/#devkit)
- convert label images into gray scale images having values between 0 and 21 as pixel values (depending on label) and 22 if its the background
- due to gpu size (max 4GB) I had to scale and pad the images to a size of 150x150

2. Convert into lmdb database
- one db for the images (Channel: 3) and one for the labels (Channel: 1)
- test size: 10; train size: 500;

3. Train the network 
- took the prototxt of segmentationcrfasrnn example and changed the lmdb database and ignore label value of the loss layer
- additionally added an accuracy layer in the end
'layer {   

  name: "accuracy"
  type: "Accuracy"
  bottom: "pred"
  bottom: "label"
  top: "accuracy"
  accuracy_param {
    ignore_label: 21
  }'
- in the solver file I only changed the test_interval to 1000 and display 200 in order to get more output
- executed command: caffe train -solver TVG_CRFRNN_new_solver.prototxt -weights TVG_CRFRNN_COCO_VOC.caffemodel

**Comment by bittnt** [comment]
Hey, glad to see that you kick things starting. 

I think you need to tune hyper-parameters a bit more, then things would work. 

Before you train CRF_RNN with FCN end-to-end, you might need to consider to make sure that you have a working FCN network. Training everything end-to-end would improve better accuracy when you set up the learning rate correctly (e.g. there are a lot of learning rates, base_lr, lr for each layer, and etc).

**Comment by annkitkat** [comment]
Thanks :) I'll try that

**Comment by annkitkat** [comment]
I downloaded the caffemodel for the FCN weights from http://www.robots.ox.ac.uk/~szheng/Res_CRFRNN/TVG_CRFRNN_COCO_VOC.caffemodel and used it for initializing the end-to-end net. Isn't this omitting the step of first training the FCN on its own?

**Comment by netw0rkf10w** [comment]
@bittnt Would you mind sharing the pretrained weights of the FCN-8s part that was used to train end-to-end your CRFasRNN?
